### ISSUE TITLE
API to allow annotations application to type arguments.

### ISSUE THREAD

**Issue by oldergod** [square/moshi#666]
So I can define such things like `List<@Qualifier Long>`.

Current APIs allow qualifier to apply to the class only. The code will apply `@Qualifier` to the type `List`. There is no way to apply `@Qualifier` to the argument `Long` of the type here.
```
Type type = Types.newParameterizedType(List.class, Long.class);
Class<? extends Annotation>[] qualifiers = new Class[] {Qualifier.class};
moshi.adapter(fieldType, qualifiers);
```

**Comment by JakeWharton** [comment]
If you have a correctly-annotated `Type` instance for `@Qualifer Long` then `Types.newParameterizedType` will do the correct thing and produce `List<@Qualifier Long>`. Whether Moshi should be in the business of creating annotated types or not is the real question here.

**Comment by swankjesse** [comment]
Moshi doesn’t currently honor annotations when types themselves implement `AnnotatedElement`.

**Comment by JakeWharton** [comment]
Equals doesn't handle this?!?

**Comment by swankjesse** [comment]
We use `Util.canonicalize()` to get an instance of `java.lang.reflect.Type` that implements `equals()` in a reasonable way, and along the way we inadvertently strip the annotations.

We don’t really want `equals()` here either. If we had that, then an adapter for `Taco` wouldn‘t match `@Nonnull Taco` and that would be bad.

**Comment by swankjesse** [comment]
Yeah, we just gotta honor `@JsonQualifier` however you get it to Moshi. Right now we do it “Java 5” style where we don’t ask types whether they are intrinsically annotated.

**Comment by swankjesse** [comment]
Looks like it’ll be quite difficult to fix this on Android because the corresponding metadata is just absent.

**Comment by oldergod** [comment]
Memo: we'd need that so we can understand things like `List<@Uint64 Long>` without having to write a adapter for this list but just relying on moshi to use the right adapter for each element.
